File: 1_카카오/프로그래머스lv2-문자열 압축.js
// unit: number, s: string => string[]
function divideString(unit, s) {
  let array = []
  for (let i = 0; i < s.length; i += unit) {
    array.push(s.slice(i, i + unit))
  }
  return array
}

function solution(s) {
  let result = s.length
  for (let unit = 1; unit < Math.floor(s.length / 2) + 1; unit++) {
    const array = divideString(unit, s)
    let temp = ''
    let count = 1
    let cur = array[0]
    for (let i = 1; i < array.length; i++) {
      if (cur === array[i]) {
        count++
      } else {
        if (count > 1) temp += String(count)
        temp += cur
        cur = array[i]
        count = 1
      }
    }
    if (count > 1) temp += String(count)
    temp += cur
    if (result > temp.length) {
      result = temp.length
    }
  }
  return result
}

// 정확성  테스트
// 테스트 1 〉	통과 (0.17ms, 30.2MB)
// 테스트 2 〉	통과 (0.30ms, 30.1MB)
// 테스트 3 〉	통과 (0.21ms, 30.4MB)
// 테스트 4 〉	통과 (0.20ms, 30.3MB)
// 테스트 5 〉	통과 (0.06ms, 30.3MB)
// 테스트 6 〉	통과 (0.13ms, 30.4MB)
// 테스트 7 〉	통과 (0.33ms, 30.4MB)
// 테스트 8 〉	통과 (0.34ms, 30.3MB)
// 테스트 9 〉	통과 (0.49ms, 30.3MB)
// 테스트 10 〉	통과 (3.18ms, 32.6MB)
// 테스트 11 〉	통과 (0.18ms, 30MB)
// 테스트 12 〉	통과 (0.19ms, 30.3MB)
// 테스트 13 〉	통과 (0.19ms, 30.4MB)
// 테스트 14 〉	통과 (0.42ms, 30.2MB)
// 테스트 15 〉	통과 (0.16ms, 30.2MB)
// 테스트 16 〉	통과 (0.12ms, 30.1MB)
// 테스트 17 〉	통과 (1.29ms, 32.9MB)
// 테스트 18 〉	통과 (1.32ms, 33.1MB)
// 테스트 19 〉	통과 (1.37ms, 32.8MB)
// 테스트 20 〉	통과 (2.01ms, 32.8MB)
// 테스트 21 〉	통과 (1.94ms, 32.8MB)
// 테스트 22 〉	통과 (1.96ms, 32.7MB)
// 테스트 23 〉	통과 (1.97ms, 33MB)
// 테스트 24 〉	통과 (1.89ms, 32.7MB)
// 테스트 25 〉	통과 (1.94ms, 32.2MB)
// 테스트 26 〉	통과 (1.91ms, 32.9MB)
// 테스트 27 〉	통과 (2.00ms, 32.8MB)
// 테스트 28 〉	통과 (0.13ms, 30.2MB)

function solution(s) {
  const N = s.length
  const LIMIT = Math.floor(N / 2)

  let result = N

  for (let unit = 1; unit < LIMIT + 1; unit++) {
    const stack = []
    let count = 1
    for (let index = 0; index < N; index += unit) {
      const top = stack[stack.length - 1]
      const cur = s.slice(index, index + unit)
      if (top === cur) {
        count++
      } else {
        if (count > 1) {
          stack[stack.length - 1] = String(count) + top
          count = 1
        }
        stack.push(cur)
      }
    }
    if (count > 1) {
      stack[stack.length - 1] = String(count) + stack[stack.length - 1]
    }
    const compression = stack.join('')
    result = Math.min(result, compression.length)
  }
  return result
}
// 정확성  테스트
// 테스트 1 〉	통과 (0.22ms, 33.5MB)
// 테스트 2 〉	통과 (0.38ms, 33.6MB)
// 테스트 3 〉	통과 (0.27ms, 33.5MB)
// 테스트 4 〉	통과 (0.19ms, 33.4MB)
// 테스트 5 〉	통과 (0.07ms, 33.4MB)
// 테스트 6 〉	통과 (0.19ms, 33.4MB)
// 테스트 7 〉	통과 (0.42ms, 33.5MB)
// 테스트 8 〉	통과 (0.44ms, 33.6MB)
// 테스트 9 〉	통과 (0.52ms, 33.6MB)
// 테스트 10 〉	통과 (1.62ms, 34.3MB)
// 테스트 11 〉	통과 (0.22ms, 33.5MB)
// 테스트 12 〉	통과 (0.22ms, 33.4MB)
// 테스트 13 〉	통과 (0.23ms, 33.4MB)
// 테스트 14 〉	통과 (0.56ms, 33.6MB)
// 테스트 15 〉	통과 (0.23ms, 33.4MB)
// 테스트 16 〉	통과 (0.08ms, 33.4MB)
// 테스트 17 〉	통과 (0.80ms, 33.8MB)
// 테스트 18 〉	통과 (0.76ms, 33.8MB)
// 테스트 19 〉	통과 (0.78ms, 33.7MB)
// 테스트 20 〉	통과 (4.44ms, 37.5MB)
// 테스트 21 〉	통과 (4.34ms, 37.5MB)
// 테스트 22 〉	통과 (4.40ms, 37.5MB)
// 테스트 23 〉	통과 (1.57ms, 34.4MB)
// 테스트 24 〉	통과 (1.57ms, 34.3MB)
// 테스트 25 〉	통과 (4.34ms, 37.6MB)
// 테스트 26 〉	통과 (1.63ms, 34.4MB)
// 테스트 27 〉	통과 (4.59ms, 37.6MB)
// 테스트 28 〉	통과 (0.18ms, 33.4MB)

File: 1_카카오/프로그래머스lv2-문자열 압축.py
def solution(s):
    # 문자열를 단위만큼 쪼개서 list로 반환하는 함수
    def string_split(unit):
        result = []
        for i in range(0, len(s), unit):
            result.append(s[i:i + unit])
        return result
    
    # 처음 비교값은 s의 길이
    result = len(s)
    # 단위는 문자열의 절반만큼만 확인하면 된다.
    for unit in range(1, (len(s) // 2) + 1):
        arr = string_split(unit)
        std = arr[0]
        temp = ""
        count = 1
        for i in range(1, len(arr)):
            if std == arr[i]:
                count += 1
            else:
                if count > 1:
                    temp += str(count)
                    count = 1
                temp += std
                std = arr[i]
                    
        if count > 1:
            temp += str(count)
        temp += std
        
        if len(temp) < result:
            result = len(temp)
            
    return result
"""
정확성  테스트
테스트 1 〉	통과 (0.03ms, 10.2MB)
테스트 2 〉	통과 (0.37ms, 10.2MB)
테스트 3 〉	통과 (0.36ms, 10.3MB)
테스트 4 〉	통과 (0.03ms, 10.2MB)
테스트 5 〉	통과 (0.00ms, 10.3MB)
테스트 6 〉	통과 (0.08ms, 10.3MB)
테스트 7 〉	통과 (0.75ms, 10.4MB)
테스트 8 〉	통과 (0.42ms, 10.3MB)
테스트 9 〉	통과 (0.59ms, 10.2MB)
테스트 10 〉	통과 (2.41ms, 10.2MB)
테스트 11 〉	통과 (0.12ms, 10.3MB)
테스트 12 〉	통과 (0.17ms, 10.3MB)
테스트 13 〉	통과 (0.11ms, 10.2MB)
테스트 14 〉	통과 (0.58ms, 10.2MB)
테스트 15 〉	통과 (0.11ms, 10.4MB)
테스트 16 〉	통과 (0.02ms, 10.3MB)
테스트 17 〉	통과 (1.05ms, 10.3MB)
테스트 18 〉	통과 (1.02ms, 10.3MB)
테스트 19 〉	통과 (1.13ms, 10.3MB)
테스트 20 〉	통과 (2.40ms, 10.3MB)
테스트 21 〉	통과 (2.56ms, 10.2MB)
테스트 22 〉	통과 (2.61ms, 10.2MB)
테스트 23 〉	통과 (2.33ms, 10.4MB)
테스트 24 〉	통과 (2.12ms, 10.3MB)
테스트 25 〉	통과 (2.44ms, 10.3MB)
테스트 26 〉	통과 (2.32ms, 10.3MB)
테스트 27 〉	통과 (2.58ms, 10.3MB)
테스트 28 〉	통과 (0.02ms, 10.4MB)
"""

# 22년 7월 풀이
# 문자열 + 스택 풀이
def solution(s):
    N = len(s)
    LIMIT = N // 2
    
    result = N
    
    for unit in range(1, LIMIT + 1):
        stack = []
        count = 1
        for index in range(0, N, unit):
            cur = s[index : index + unit]
            # 같은 문자가 들어오면 count만 센다.
            if stack and stack[-1] == cur:
                count += 1
            # 다른 문자가 들어오면...
            else:
                # 그동안 세줬던 count를 확인한다. 만약 count가 1보다 크면...
                if count > 1:
                    stack[-1] = str(count) + stack[-1]
                    count = 1
                stack.append(cur)
        # 남아있는 count 확인
        if count > 1:
            stack[-1] = str(count) + stack[-1]
        # 압축된 문자열의 길이와 비교
        compression = "".join(stack)
        result = min(result, len(compression))
    
    return result
"""
정확성  테스트
테스트 1 〉	통과 (0.03ms, 10.3MB)
테스트 2 〉	통과 (0.29ms, 10.3MB)
테스트 3 〉	통과 (0.15ms, 10.2MB)
테스트 4 〉	통과 (0.03ms, 10.3MB)
테스트 5 〉	통과 (0.00ms, 10.1MB)
테스트 6 〉	통과 (0.04ms, 10.2MB)
테스트 7 〉	통과 (0.31ms, 10.2MB)
테스트 8 〉	통과 (0.32ms, 10.3MB)
테스트 9 〉	통과 (0.45ms, 10.2MB)
테스트 10 〉	통과 (1.61ms, 10.4MB)
테스트 11 〉	통과 (0.07ms, 10.2MB)
테스트 12 〉	통과 (0.07ms, 10.3MB)
테스트 13 〉	통과 (0.09ms, 10.1MB)
테스트 14 〉	통과 (0.45ms, 10.1MB)
테스트 15 〉	통과 (0.08ms, 10.2MB)
테스트 16 〉	통과 (0.01ms, 10.2MB)
테스트 17 〉	통과 (0.82ms, 10.3MB)
테스트 18 〉	통과 (0.79ms, 10.2MB)
테스트 19 〉	통과 (0.78ms, 10.3MB)
테스트 20 〉	통과 (1.72ms, 10.2MB)
테스트 21 〉	통과 (1.73ms, 10.3MB)
테스트 22 〉	통과 (1.80ms, 10.1MB)
테스트 23 〉	통과 (1.65ms, 10.3MB)
테스트 24 〉	통과 (1.56ms, 10.1MB)
테스트 25 〉	통과 (1.73ms, 10.1MB)
테스트 26 〉	통과 (1.72ms, 10.1MB)
테스트 27 〉	통과 (1.72ms, 10.2MB)
테스트 28 〉	통과 (0.02ms, 10.2MB)
"""